Userland/Libraries/LibJIT/Assembler.h

```diff
@@ -117,89 +122,201 @@ struct Assembler {
+                u8 rm : 3;
+                u8 reg : 3;
+                u8 mode : 2;
+            };
+            u8 raw;
+        };
+
+        static void emit_slash(Assembler& assembler, u8 slash, Operand rm, Patchable patchable = Patchable::No)
+        {
+            Raw raw;
+            raw.rm = encode_reg(rm.reg);
+            raw.reg = slash;
+            emit(assembler, raw, rm, patchable);
+        }
+
+        static void emit_rm(Assembler& assembler, Operand dst, Operand src, Patchable patchable = Patchable::No)
+        {
+            VERIFY(dst.type == Operand::Type::Reg);
+            Raw raw {};
+            raw.reg = encode_reg(dst.reg);
+            raw.rm = encode_reg(src.reg);
+            emit(assembler, raw, src, patchable);
+        }
+
+        static void emit_mr(Assembler& assembler, Operand dst, Operand src, Patchable patchable = Patchable::No)
+        {
+            VERIFY(src.type == Operand::Type::Reg);
+            Raw raw {};
+            raw.reg = encode_reg(src.reg);
+            raw.rm = encode_reg(dst.reg);
+            emit(assembler, raw, dst, patchable);
+        }
+
+    private:
+        static void emit(Assembler& assembler, Raw raw, Operand rm, Patchable patchable)
+        {
+            // FIXME: rm:100 (RSP) is reserved as the SIB marker
+            VERIFY(rm.type != Operand::Type::Imm);
+
+            switch (rm.type) {
+            case Operand::Type::Reg:
+                // FIXME: There is mod:00,rm:101(EBP?) -> disp32, that might be something else
+                raw.mode = Reg;
+                assembler.emit8(raw.raw);
+                break;
+            case Operand::Type::Mem64BaseAndOffset: {
+                auto disp = rm.offset_or_immediate;
+                if (patchable == Patchable::Yes) {
+                    raw.mode = MemDisp32;
+                    assembler.emit8(raw.raw);
+                    assembler.emit32(disp);
+                } else if (disp == 0) {
+                    raw.mode = Mem;
+                    assembler.emit8(raw.raw);
+                } else if (static_cast<i64>(disp) >= -128 && disp <= 127) {
+                    raw.mode = MemDisp8;
+                    assembler.emit8(raw.raw);
+                    assembler.emit8(disp & 0xff);
+                } else {
+                    raw.mode = MemDisp32;
+                    assembler.emit8(raw.raw);
+                    assembler.emit32(disp);
+                }
+                break;
+            }
+            case Operand::Type::Imm:
+                VERIFY_NOT_REACHED();
+            }
+        }
+    };
+
+    struct REX {
+        union Raw {
+            struct {
+                u8 B : 1; // ModRM::RM
+                u8 X : 1; // SIB::Index
+                u8 R : 1; // ModRM::Reg
+                u8 W : 1; // Operand size override
+                u8 _ : 4 { 0b0100 };
+            };
+            u8 raw;
+        };
+        static void emit_for_OI(Assembler& assembler, Operand arg, bool W)
```

Comment:
Why the arbitrary coding style violation to use uppercase on `OI` and `W` here? (And elsewhere)
Also what *is* `OI`?

Reply:
These are spec names the field in the REX prefix is just called `W`
and as for OI, this is what the spec uses to describe the addressing mode of some instructions
I think it might stand for Opcode+Immediate, but intel is not very good in telling stuff like this